### ISSUE TITLE
Sometimes there is a delay of several minutes before the shadow delta callback is invoked.

### ISSUE THREAD

**Issue by mrducky4** [aws/aws-iot-device-sdk-python#244]
When some other entity changes the desired shadow document state, we expect the shadow delta callback on our device to be called immediately. Usually that is the case, but sometimes there is a delay of several minutes. When we measured the delay, in all cases it was either about 2.5 minutes or about 4.5 minutes.

I set up an IoT rule to send shadow document messages to an SQS queue. In the case when the delay occurs, I can see in the queue that the message for "$aws/things/myThingName/shadow/update" appeared in the queue immediately after changing the desired state.

We cannot always reproduce this delay. But when it occurs, it seems to occur in groups. So if it happens once, it is more likely to happen again in the near future.

Any ideas what can cause this, and what we can do to fix it?

```
clientID = "myThingName"

iotClient = AWSIoTMQTTClient(clientID=clientID, useWebsocket=False)
iotClient.configureEndpoint("xxxxxx-ats.iot.us-east-1.amazonaws.com", 8883)
iotClient.configureCredentials(rootCA, key, cert)

# sample configuration from example files
iotClient.configureAutoReconnectBackoffTime(1, 32, 20)
iotClient.configureConnectDisconnectTimeout(10)  # 10 sec
iotClient.configureMQTTOperationTimeout(5)  # 5 sec

iotClient.connect()

shadowbroker = AWSIoTMQTTShadowClient(clientID, awsIoTMQTTClient=iotClient)
shadowHandler = shadowbroker.createShadowHandlerWithName(clientID, True)
shadowHandler.shadowRegisterDeltaCallback(shadowDeltaHandler)

# shadowDeltaHandler is a callback that sometimes is not called until minutes after we expect
```

The same code also handles job requests as in jobsSample.py.

**Comment by bretambrose** [comment]
Do you have logs from your application that cover a period of time that includes this behavior?  I'm wondering if the application is suffering an mqtt disconnect during the interval.  On what kind of platform is the application running?

**Comment by mrducky4** [comment]
We get logs like this when making a change to the shadow document state:
```
2020-03-09 18:32:04,005 Thread-2 DEBUG AWSIoTPythonSDK.core.protocol.internal.workers: Produced [message] event
2020-03-09 18:32:04,006 Thread-1 DEBUG AWSIoTPythonSDK.core.protocol.internal.workers: Dispatching [message] event
2020-03-09 18:32:04,012 Thread-1 DEBUG AWSIoTPythonSDK.core.protocol.internal.clients: Invoking custom event callback...
2020-03-09 18:32:04,015 Thread-2 DEBUG AWSIoTPythonSDK.core.protocol.internal.workers: Produced [message] event
2020-03-09 18:32:04,015 Thread-1 DEBUG AWSIoTPythonSDK.core.protocol.internal.workers: Dispatching [message] event
shadowDeltaHandler() {"version":267,"timestamp":1583793123,"state":{"mode":"IDLE"},"metadata":{"mode":{"timestamp":1583793123}}}
2020-03-09 18:32:04,021 Thread-1 DEBUG AWSIoTPythonSDK.core.protocol.internal.clients: Invoking custom event callback...
2020-03-09 18:32:04,022 Thread-379 INFO AWSIoTPythonSDK.core.protocol.mqtt_core: Performing sync publish...
2020-03-09 18:32:04,025 Thread-2 DEBUG AWSIoTPythonSDK.core.protocol.internal.workers: Produced [puback] event
2020-03-09 18:32:04,026 Thread-1 DEBUG AWSIoTPythonSDK.core.protocol.internal.workers: Dispatching [puback] event
2020-03-09 18:32:04,087 Thread-2 DEBUG AWSIoTPythonSDK.core.protocol.internal.workers: Produced [message] event
2020-03-09 18:32:04,088 Thread-1 DEBUG AWSIoTPythonSDK.core.protocol.internal.workers: Dispatching [message] event
2020-03-09 18:32:04,090 Thread-1 DEBUG AWSIoTPythonSDK.core.shadow.deviceShadow: shadow message clientToken: 1c63ca57-b742-4237-92e1-c3154068a3a5
2020-03-09 18:32:04,090 Thread-1 DEBUG AWSIoTPythonSDK.core.shadow.deviceShadow: Token is in the pool. Type: accepted
2020-03-09 18:32:04,092 Thread-1 DEBUG AWSIoTPythonSDK.core.protocol.internal.clients: Invoking custom event callback...
```
When the delay occurs, there are no other log messages. These normal messages appear after the delay. I have seen the delay happen while I was actively using an ssh connection to the device. There was no indication of any network trouble.

I'm trying to reproduce the problem while capturing packets with wireshark or tcpdump, but haven't reproduced it yet.

The real device runs on nvidia jetson nano with linux. But we have also seen the problem with the same IoT device code running on Windows, as a test.

**Comment by bretambrose** [comment]
That looks like logs from the shadow state setter.  Do you have matching logs from the application that is supposed to receive the update (for the length of the delayed receive)?

**Comment by mrducky4** [comment]
That is the log from the application that is supposed to receive the update. Those are the messages that show the device receiving the update, the SDK calls back to our shadowDeltaHandler() function, then we handle the state change request and set a new shadow state with a reported state to match the requested desired state we just received, and a new desired state of null.

The log from our device during the length of the delayed receive looks exactly like that, with no activity in the log during the delay period, then all those log messages appear after the delay.

Next time I can reproduce the issue I will post an update here with a new log.

**Comment by mrducky4** [comment]
I have more data about another occurence of this delay. I also opened an AWS support case about this.
When some other entity changes the desired shadow document state, we expect the shadow delta callback on our device to be called immediately. Usually that is the case, but sometimes there is a delay of several minutes. The same thing happens when submitting a new IoT job for the device.

Our device uses the IoT device SDK for python, running on a jetson nano board. But we have also seen the same symptom when running a test version of our device, with the same python sdk, on Windows.

I have details and logs for a specific case of a string of delays while updating the shadow document, occuring on 2020-03-18. Times are in GMT-4.
11:32:03 change shadow desired state
11:36:08 device gets notification of the new desired state. 4:03 delay.

13:07:58 change shadow desired state
13:12:05 device gets notification of the new desired state. 4:07 delay.

13:43:03 change shadow desired state
13:47:12 device gets notification of the new desired state. 4:09 delay.

I set up an IoT rule to send shadow document messages to an SQS queue. In the case when the delay occurs, I can see in the queue that the message for "$aws/things/myThingName/shadow/update" appeared in the queue immediately after changing the desired state.

I know the device's network connection stays up the whole time.

I have a pcap file captured with tcpdump on the device. Observations from the pcap file:
A few application data packets occur every 10 minutes until 11:24:09
11:32:03 - no packets at all
11:34:10 - application data and "previous segment not captured" warning
         - this is the time to expect the 10 minute periodic activity
11:36:08 - retransmission warning
11:36:08 - application data packets, consistent with communication recovering

The same pattern repeats for the other cases of delay.

We cannot always reproduce this delay. But when it occurs, it seems to occur in groups. So if it happens once, it is more likely to happen again in the near future.

Attachment includes the pcap file and the device log from our app, including messages from the python IoT device SDK.
[iot_delay.zip](https://github.com/aws/aws-iot-device-sdk-python/files/4359630/iot_delay.zip)
